plugins/modules/mso_schema_template_service_graph.py

```diff
@@ -0,0 +1,264 @@
+    index = 0
+    Nodes = []
+
+    # Get service graphs
+    service_graphs = [f.get('name') for f in schema_obj.get('templates')[template_idx]['serviceGraphs']]
+    if service_graph_name in service_graphs:
+        service_graph_idx = service_graphs.index(service_graph_name)
+    
+    # Get service nodes
+    query_node_data = mso.query_nodes()
+    service_nodes = [f.get('name') for f in query_node_data]
+    if service_node_list is not None:
+      for node in service_node_list:
+        node_name = node.get('type')
+        if node_name in service_nodes:
+          index = index + 1
+          for node_data in query_node_data:
+            if node_data['name'] == node_name:
+              node_data = node_data
+          Nodes.append(dict(
+              name=node_name,
+              serviceNodeTypeId=node_data.get('id'),
+              index = index,
+              serviceNodeRef=dict(   
+                serviceNodeName=node_name,
+                serviceGraphName=service_graph_name,
+                templateName=template,
+                schemaId=schema_id,
+          )),
+          )
+  
+    if state == 'query':
+      if service_graph_name is None:
+            mso.existing = schema_obj.get('templates')[template_idx]['serviceGraphs']
+      if service_graph_name is not None and service_graph_idx is None:
+            mso.fail_json(msg="service_graph '{service_graph}' not found".format(service_graph=service_graph_name))
+      elif service_graph_idx is not None:
+            mso.existing = schema_obj.get('templates')[template_idx]['serviceGraphs'][service_graph_idx]
+      mso.exit_json()
+
+    service_graphs_path = '/templates/{0}/serviceGraphs'.format(template)
+    service_graph_path = '/templates/{0}/serviceGraphs/{1}'.format(template, service_graph_name)
+    service_nodes_path = '/templates/{0}/serviceGraphs/{1}/serviceNodes'.format(template, service_graph_name)
+    ops = []
+
+    mso.previous = mso.existing
+    if state == 'absent':
+        mso.proposed = mso.sent = {}
+        if service_graph_idx is None:
+            # There was no service graph to begin with
+            pass
+        else:
+            mso.existing = {}
+            ops.append(dict(op='remove', path=service_graph_path))
+
+    elif state == 'present':
+        mso.sanitize(Nodes, collate=True)
+        if service_graph_idx is None:
+            # service graph does not exist, so we have to create it
+            if service_graph_display_name is None:
+                service_graph_display_name = service_graph_name
+
+            payload = dict(
+                name=service_graph_name,
+                displayName=service_graph_display_name,
+                description=service_graph_description,
+                nodeFilter=node_filter,
+                serviceGraphRef=dict(
+                  serviceGraphName=service_graph_name,
+                  templateName=template,
+                  schemaId=schema_id,
+            ),
```

Comment:
this should be more on the right missing spaces.

Reply:
ok